store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/util/ConcatenateMergeFunction.java

```diff
@@ -16,27 +16,41 @@
 
 package uk.gov.gchq.gaffer.federatedstore.util;
 
-import com.google.common.collect.Lists;
 import org.apache.commons.lang3.builder.HashCodeBuilder;
 
-import uk.gov.gchq.koryphe.impl.function.IterableConcat;
-import uk.gov.gchq.koryphe.impl.function.ToList;
+import uk.gov.gchq.koryphe.iterable.ChainedIterable;
 
 import java.util.Collections;
 import java.util.function.BiFunction;
 
-import static java.util.Objects.isNull;
 import static java.util.Objects.nonNull;
 
-public class ConcatenateListMergeFunction implements BiFunction<Object, Iterable<Object>, Iterable<Object>> {
+/**
+ * A BiFunction for concatenating results from multiple graphs into one Iterable.
+ */
+public class ConcatenateMergeFunction implements BiFunction<Object, Iterable<Object>, Iterable<Object>> {
 
+    /**
+     * Concatenates the result from a graph with the previous results Iterable.
+     *
+     * @param update    the result to be added
+     * @param state     the iterable of previous results
+     */
     @Override
     public Iterable<Object> apply(final Object update, final Iterable<Object> state) {
-        //When update=null, Then this STOPS ToList returning an Iterable with a null in it.
-        final Iterable<Object> updateSafe = isNull(update) ? Collections.emptyList() : (Iterable<Object>) new ToList().apply(update);
-        return isNull(state)
-                ? updateSafe
-                : new IterableConcat<>().apply(Lists.newArrayList(updateSafe, state));
+        final Iterable<Object> updateSafe;
+        // When update is null, don't add an Iterable containing a null.
+        if (update == null) {
+            updateSafe = Collections.emptyList();
+        } else if (!(update instanceof Iterable)) {
+            // When update is not Iterable, wrap it in a List.
+            updateSafe = Collections.singletonList(update);
```

Comment:
So an array becomes an Iterable with one object of type array in it?
`Iterable{[1,2,3]}`

I think the old would return the contents of the array as the iterable.
`Iterable{1,2,3}`

Reply:
See: `ToList` line 44

Reply:
If by the old one, you mean the way it was done prior to the introduction of this class, then yes. @t92549 and I had a look at this and the old merge did not convert a result to an iterable if it wasn't one already ([see this code](https://github.com/gchq/Gaffer/blob/9421d9dfed70cfa14c475b8f8d9144589bc135ab/store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedOperationChainHandler.java#L66)).

That behaviour changed with the introduction of this class. The current behaviour seems better.

Old behaviour with results from a graph A & B, where one graph doesn't return an `Iterable` for some reason:
`A -> Array [1,2,3]` _not an `Iterable`_
`B -> Iterable {4,5,6}`
When merged returns `Iterable{[1,2,3],{4,5,6}}`, these are not actually merged because a non `Iterable` was found (see the `break` in the for loop).

New (current) behaviour with those same results returns `Iterable{[1,2,3],4,5,6}`. The array is still an array, but the second set of results have been merged better.

An example where the current behaviour is more obviously better is this:
`A -> Int 1` _not an `Iterable`_
`B -> Iterable {4,5,6}`
When merged returns `Iterable{1,4,5,6}`, old way would have given `Iterable{1,{4,5,6}}`.

This is why it was considered useful to add tests, although they don't seem to cover any of those use cases yet.

We also thought it was uncommon for results to not be an `Iterable`, so these differences are unlikely to be noticed.

Reply:
I'm talking in reference to class your replacing `ConcatenateListMergeFunction`

```java
final String[] s = new String[]{"one","two","three"};
//        final Iterable<Object> apply = new ConcatenateListMergeFunction().apply(s, null);
        final Iterable<Object> apply = new ConcatenateMergeFunction().apply(s, null);
        org.assertj.core.api.Assertions.assertThat(apply)
                .containsExactly("one","two","three");
```
                
Toggle between ConcatenateListMergeFunction (passes) and ConcatenateMergeFunction (fails).

Reply:
I have another look and you're correct, Arrays are now handled differently. I'll fix this. The way they were handled before this class was introduced was also different.

Reply:
Added a test to verify it will work how @GCHQDev404 wants.
This will mean it behaves differently to Gaffer 1 though so perhaps another ticket should be made to compare before and after?